p2p/src/authenticated/discovery/mod.rs

```diff
@@ -324,11 +327,17 @@ mod tests {
                             // Add to received set
                             received.insert(sender);
                         }
+                        complete_sender.send(()).await.unwrap();
+
+                        // Process messages until all finished (or else sender loops could get stuck as a peer may drop)
+                        loop {
```

Comment:
It is possible that finished receivers were causing the runtime to get stuck (if too many peers weren't reading on their channel, would cause blocking).